From 3443ad0be2bca470f7dceb45efc3f626c28ec59c Mon Sep 17 00:00:00 2001
From: Simon Busch <morphis@gravedo.de>
Date: Mon, 5 Aug 2013 16:53:42 +0000
Subject: [PATCH 3/3] Handle connection error by exiting the application

If the compositor connection dies, it is an unrecoverable error.

Upstream-Status: Pending (https://bugreports.qt-project.org/browse/QTBUG-29001)
---
 src/plugins/platforms/wayland_common/qwaylanddisplay.cpp     | 12 ++++++++++--
 src/plugins/platforms/wayland_common/qwaylandeventthread.cpp |  5 ++++-
 2 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/src/plugins/platforms/wayland_common/qwaylanddisplay.cpp b/src/plugins/platforms/wayland_common/qwaylanddisplay.cpp
index 4586ff8..d56bf5d 100644
--- a/src/plugins/platforms/wayland_common/qwaylanddisplay.cpp
+++ b/src/plugins/platforms/wayland_common/qwaylanddisplay.cpp
@@ -162,13 +162,21 @@ QWaylandDisplay::~QWaylandDisplay(void)
 
 void QWaylandDisplay::flushRequests()
 {
-    wl_display_dispatch_queue_pending(mDisplay, mEventQueue);
+    int ret = wl_display_dispatch_queue_pending(mDisplay, mEventQueue);
+    if (ret < 0) {
+        qWarning("The wayland connection broke (error %d). Did the wayland compositor die?", errno);
+        exit(1);
+    }
     wl_display_flush(mDisplay);
 }
 
 void QWaylandDisplay::blockingReadEvents()
 {
-    wl_display_dispatch_queue(mDisplay, mEventQueue);
+    int ret = wl_display_dispatch_queue(mDisplay, mEventQueue);
+    if (ret < 0) {
+        qWarning("The wayland connection broke (error %d). Did the wayland compositor die?", errno);
+        exit(1);
+    }
 }
 
 QWaylandScreen *QWaylandDisplay::screenForOutput(struct wl_output *output) const
diff --git a/src/plugins/platforms/wayland_common/qwaylandeventthread.cpp b/src/plugins/platforms/wayland_common/qwaylandeventthread.cpp
index 17fa744..2e2c227 100644
--- a/src/plugins/platforms/wayland_common/qwaylandeventthread.cpp
+++ b/src/plugins/platforms/wayland_common/qwaylandeventthread.cpp
@@ -31,7 +31,10 @@ void QWaylandEventThread::displayConnect()
 
 void QWaylandEventThread::readWaylandEvents()
 {
-    wl_display_dispatch(m_display);
+    if (wl_display_dispatch(m_display) < 0) {
+        qWarning("The wayland connection broke (error %d). Did the wayland compositor die?", errno);
+        exit(1);
+    }
     emit newEventsRead();
 }
 
-- 
1.8.1.2

